mobly/controllers/android_device_lib/snippet_client_v2.py

```diff
@@ -0,0 +1,303 @@
+  def user_id(self):
+    """The user id to use for this snippet client.
+
+    All the operations of the snippet client should be used for a particular
+    user. For more details, see the Android documentation of testing
+    multiple users.
+
+    Thus this value is cached and, once set, does not change through the
+    lifecycle of this snippet client object. This caching also reduces the
+    number of adb calls needed.
+
+    Returns:
+      An integer of the user id.
+    """
+    if self._user_id is None:
+      self._user_id = self._adb.current_user_id
+    return self._user_id
+
+  def before_starting_server(self):
+    """Prepares for starting the server.
+
+    Raises:
+      errors.ServerStartPreCheckError: if the server app is not installed
+        for the current user.
+    """
+    self._check_snippet_app_installed()
+    self._disable_hidden_api_blacklist()
+
+  def _check_snippet_app_installed(self):
+    """Checks that the Mobly Snippet app is installed on the remote device.
+
+    Raises:
+      errors.ServerStartPreCheckError: if the server app is not installed
+        for the current user.
+    """
+    out = self._adb.shell(f'pm list package --user {self.user_id}')
+    if not utils.grep(f'^package:{self.package}$', out):
+      raise errors.ServerStartPreCheckError(
+          self._device,
+          f'{self.package} is not installed for user {self.user_id}.')
+
+    # Check that the app is instrumented.
+    out = self._adb.shell('pm list instrumentation')
+    matched_out = utils.grep(
+        f'^instrumentation:{self.package}/{_INSTRUMENTATION_RUNNER_PACKAGE}',
+        out)
+    if not matched_out:
+      raise errors.ServerStartPreCheckError(
+          self._device,
+          f'{self.package} is installed, but it is not instrumented.')
+
+    # Check that the instrumentation target is installed if it's not the
+    # same as the snippet package.
+    match = re.search(r'^instrumentation:(.*)\/(.*) \(target=(.*)\)$',
+                      matched_out[0])
+    target_name = match.group(3)
+    if target_name != self.package:
+      out = self._adb.shell(f'pm list package --user {self.user_id}')
+      if not utils.grep(f'^package:{target_name}$', out):
+        raise errors.ServerStartPreCheckError(
+            self._device,
+            f'Instrumentation target {target_name} is not installed for user '
+            f'{self.user_id}.')
+
+  def _disable_hidden_api_blacklist(self):
+    """If necessary and possible, disables hidden api blacklist."""
+    # Check version_codename in addition to sdk_version because Android P builds
+    # in development report sdk_version 27, but still enforce the blacklist.
+    if self._device.is_rootable and (
+        int(self._device.build_info['build_version_sdk']) >= 28 or
+        self._device.build_info['build_version_codename'] == 'P'):
+      self._device.adb.shell(
+          'settings put global hidden_api_blacklist_exemptions "*"')
+
+  def do_start_server(self):
```

Comment:
It's because we have the prefix "before_" and "after_" in stage names, but we can't use "start_server" directly because it conflicts with the high-level process name, so we add the "do_" prefix. 

Current situation: 
In the function `start_server`, there are four stages:
* `before_starting_server`
* `do_start_server`
* `build_connection`
* `after_starting_server`

Or do you think it's better to rename it to `start_remote_server` to make it clearer?

Reply:
why does it conflict with high-level stuff?
shouldn't the inheritance model handle the conflict?

Reply:
Sorry, for high-level process name, I mean the function named `start_server` has those four stages, so the second stage can't use the name `start_server` again. They are both functions defined in client base.

Reply:
Resolved this comment by #806. Rebased master.